test/laplace.jl

```diff
@@ -0,0 +1,127 @@
+function generate_data()
+    Random.seed!(1)
+    X = range(0, 23.5; length=48)
+    fs = @. 3 * sin(10 + 0.6X) + sin(0.1X) - 1
+    # invlink = normcdf
+    invlink = logistic
+    ps = invlink.(fs)
+    Y = [rand(Bernoulli(p)) for p in ps]
+    return X, Y
+end
+
+dist_y_given_f(f) = Bernoulli(logistic(f))
+
+function build_latent_gp(theta)
+    variance = softplus(theta[1])
+    lengthscale = softplus(theta[2])
+    kernel = variance * with_lengthscale(SqExponentialKernel(), lengthscale)
+    return LatentGP(GP(kernel), dist_y_given_f, 1e-8)
+end
+
+@testset "Gaussian" begin
+    # should check for convergence in one step, and agreement with exact GPR
+end
+
+@testset "gradients" begin
+    X, Y = generate_data()
+    @testset "laplace_lml" begin
+        theta0 = rand(2)
+        function objective(theta)
+            lf = build_latent_gp(theta)
+            lml = ApproximateGPs.laplace_lml(lf(X), Y)
+            return -lml
+        end
+        fd_grad = only(FiniteDifferences.grad(central_fdm(5, 1), objective, theta0))
+        ad_grad = only(Zygote.gradient(objective, theta0))
+        @test ad_grad ≈ fd_grad
+    end
+end
+
+@testset "chainrule" begin
+    xs = [0.2, 0.3, 0.7]
+    ys = [1, 1, 0]
+    K = kernelmatrix(with_lengthscale(Matern52Kernel(), 0.3), xs)
+    test_frule(
+        ApproximateGPs.newton_inner_loop,
+        dist_y_given_f,
+        ys,
+        K;
+        fkwargs=(; f_init=zero(ys), maxiter=100),
+        rtol=0.01,
+    )
+    #test_rrule(ApproximateGPs.newton_inner_loop, dist_y_given_f, ys, K; fkwargs=(;f_init=zero(ys), maxiter=100), rtol=0.05)  # my rrule might still be broken
```

Comment:
The issue may be that `K` has to be a positive-definite matrix, but the generated tangents are likely to make it non-posdef! It'd be nice if I could test `closure(A) = newton_inner_loop(dist_y_given_f, ys, A'A)` instead, but not sure how that would work ... would have to take the rrule for matmul and figure out how to call back into AD to actually call the rrule for newton_inner_loop?

Reply:
yep, fixed by 2d43113